backend/src/main/java/com/htecgroup/skynest/service/impl/FileServiceImpl.java

```diff
@@ -0,0 +1,169 @@
+import com.htecgroup.skynest.model.response.FileResponse;
+import com.htecgroup.skynest.repository.BucketRepository;
+import com.htecgroup.skynest.repository.FileMetadataRepository;
+import com.htecgroup.skynest.repository.UserRepository;
+import com.htecgroup.skynest.service.FileService;
+import com.mongodb.MongoException;
+import com.mongodb.client.gridfs.model.GridFSFile;
+import lombok.RequiredArgsConstructor;
+import org.modelmapper.ModelMapper;
+import org.springframework.core.io.InputStreamResource;
+import org.springframework.core.io.Resource;
+import org.springframework.data.mongodb.core.query.Criteria;
+import org.springframework.data.mongodb.core.query.Query;
+import org.springframework.data.mongodb.gridfs.GridFsOperations;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.web.multipart.MultipartFile;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.UUID;
+
+@Service
+@RequiredArgsConstructor
+public class FileServiceImpl implements FileService {
+
+  private final GridFsOperations operations;
+  private final BucketRepository bucketRepository;
+
+  private final CurrentUserServiceImpl currentUserService;
+  private final UserRepository userRepository;
+
+  private final FileMetadataRepository fileMetadataRepository;
+
+  private final ModelMapper modelMapper;
+
+  @Override
+  @Transactional(rollbackFor = Exception.class)
+  public FileResponse uploadFile(MultipartFile multipartFile, UUID bucketId) {
+
+    FileMetadataEntity emptyFileMetadata = reserveFileName(multipartFile, bucketId);
+
+    FileMetadataEntity savedFileMetadata = storeFileContents(multipartFile, emptyFileMetadata);
+
+    return modelMapper.map(savedFileMetadata, FileResponse.class);
+  }
+
+  private FileMetadataEntity storeFileContents(
+      MultipartFile multipartFile, FileMetadataEntity reservedFileMetadataEntity) {
+    try {
+      String fileContentId = storeFileContentsToDatabase(multipartFile);
+      reservedFileMetadataEntity.setContentId(fileContentId);
+      return fileMetadataRepository.save(reservedFileMetadataEntity);
+    } catch (IOException | MongoException e) {
+      fileMetadataRepository.delete(reservedFileMetadataEntity);
+      throw new FileIOException();
+    }
+  }
+
+  private FileMetadataEntity reserveFileName(MultipartFile multipartFile, UUID bucketId) {
+
+    FileMetadataEntity fileMetadataEntity = initFileMetadata(multipartFile, bucketId);
+
+    checkCanCreateFile(fileMetadataEntity);
+
+    return fileMetadataRepository.save(fileMetadataEntity);
+  }
+
+  private String storeFileContentsToDatabase(MultipartFile multipartFile) throws IOException {
+    Object objectId =
+        operations.store(
+            multipartFile.getInputStream(),
+            multipartFile.getOriginalFilename(),
+            multipartFile.getContentType());
+    return objectId.toString();
+  }
+
```

Comment:
The logic here seems a little convoluted, lots of helper functions. Mentors would probably ask for the checks to be removed to a separate class (or a custom annotation :)), but I can live with this.

Reply:
those helper functions are supposed to make this easier to read

not sure if it worked though

Reply:
I agree with this, it's kinda hard to read it like this, but maybe it's good practice like this I don't know.

Reply:
Maybe it would be okay if the checks are somewhere else.